src/shims/posix/linux/sync.rs

```diff
@@ -0,0 +1,130 @@
+use crate::thread::Time;
+use crate::*;
+use rustc_target::abi::{Align, Size};
+use std::time::{Instant, SystemTime};
+
+/// Implementation of the SYS_futex syscall.
+pub fn futex<'tcx>(
+    this: &mut MiriEvalContext<'_, 'tcx>,
+    args: &[OpTy<'tcx, Tag>],
+    dest: PlaceTy<'tcx, Tag>,
+) -> InterpResult<'tcx> {
+    // The amount of arguments used depends on the type of futex operation.
+    // The full futex syscall takes six arguments (excluding the syscall
+    // number), which is also the maximum amount of arguments a linux syscall
+    // can take on most architectures.
+    // However, not all futex operations use all six arguments. The unused ones
+    // may or may not be left out from the `syscall()` call.
+    // Therefore we don't use `check_arg_count` here, but only check for the
+    // number of arguments to fall within a range.
+    if !(4..=7).contains(&args.len()) {
+        throw_ub_format!("incorrect number of arguments for futex syscall: got {}, expected between 4 and 7 (inclusive)", args.len());
+    }
+
+    // The first three arguments (after the syscall number itself) are the same to all futex operations:
+    //     (int *addr, int op, int val).
+    // We checked above that these definitely exist.
+    let addr = this.read_immediate(args[1])?;
+    let op = this.read_scalar(args[2])?.to_i32()?;
+    let val = this.read_scalar(args[3])?.to_i32()?;
+
+    // The raw pointer value is used to identify the mutex.
+    // Not all mutex operations actually read from this address or even require this address to exist.
+    let futex_ptr = this.force_ptr(addr.to_scalar()?)?;
```

Comment:
```suggestion
    // This will make FUTEX_WAKE fail on an integer cast to a pointer. But FUTEX_WAIT on
    // such a pointer can never work anyway, so that seems fine.
    let futex_ptr = this.force_ptr(addr.to_scalar()?)?;
```
Is this right?